### PR TITLE
Adding :class: ManimColor to manim and converting all types

### LINKED ISSUE #2837: Finally fixing the Color situation
## To fix

As discussed on discord

The color situation in Manim is pretty janked right now and there are 2-3 different ways to manage Color. We do not really need all that duplicate functionality so i think we settled now on reworking the Enum in  color.py to be the source of truth again. Making it Colors(str, Enum) to not break current functionality (but also benchmark that at first before doing that change because it could have drastic performance impact) and then so that IDE doesn't cry we add a .pyi file with the exported colors so that autocomplete works and then ditching colour.py because it is not compatible with pickle

- https://github.com/ManimCommunity/manim/issues/2451
> ## Enhancement proposal
> 
>     * Manim's colors are defined twice, [once in an enum](https://github.com/ManimCommunity
>/manim/blob/main/manim/utils/color.py#L33-L260) and once as [constants](https://github.com
>/ManimCommunity/manim/blob/main/manim/utils/color.py#L286-L366). It seems like the Enum was 
>introduced in [# New colour enums- Try2 #488](https://github.com/ManimCommunity/manim/pull/488) 
>as an attempt to clean up the code, with constants having to be explicitly imported. However, 
>for convenience the community eventually decided to import all the constants, and then to 
>appease people's editors and linters the constants eventually had to be explicitly defined as 
>well. I think we should either only define colors in an enum, or define them as constants, but 
>not both. Seeing how the natural trend is to go towards convenience, I think they should be 
>defined as constants. Furthermore, they should be instances of the `Color` class from the 
>colour module (or @marcin-serwin's fork).
> 
>     * We have many functions for [converting between color representations]
>(https://github.com/ManimCommunity/manim/blob/main/manim/utils/color.py#L453-L478) that 
>duplicate the functionality of existing, specialized libraries. Any uses of those should be 
>replaced, probably by methods of `Color`, and the functions eventually removed.
> 
> 
> ## Additional comments

Related
- https://github.com/ManimCommunity/manim/pull/2606
- #2605

### PR DESCRIPTION
**Pull request by MrDiver**
## Overview: What does this pull request change?
This pull request aims to remove all the duplicates of color handling in Manim and unify all colors under one Color type named `ManimColor` and a TypeAlias `ParsableManimColor` which represents all types that can be converted to a `ManimColor`

## Reviewer Checklist
- [x] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [x] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [x] If applicable: newly added functions and classes are tested

closes #2837

### CODE REVIEW

File: manim/utils/color.py [hunk 1]
```diff
@@ -21,350 +26,436 @@
...
+
+class ManimColor:
+    def __init__(
+        self,
+        value: str
+        | int
+        | list[int, int, int]
+        | list[int, int, int, int]
+        | list[float, float, float]
+        | list[float, float, float, float],
```
Comment (Viicos) [inline manim/utils/color.py:191]: Maybe using a ``TypeAlias` here as it is used in `ParsableManimColor` as well?

Comment (MrDiver) [reply manim/utils/color.py:191]: I guess that would be a reasonable idea, but it's more readable with autocomplete and typehints like this 🤔

**Review by behackl — COMMENTED** [review]
Is `TypeAlias` Python 3.10+ only? I'd perhaps be fine with deprecating Python 3.7, but skipping everything until 3.10 is somewhat inconvenient. :-)

**Comment by MrDiver** [comment]
I think it was more about the reuse of all the lists in ManimColor and ParsableManimColor

**Comment by Viicos** [comment]
> Is `TypeAlias` Python 3.10+ only? I'd perhaps be fine with deprecating Python 3.7, but skipping everything until 3.10 is somewhat inconvenient. :-)

Ah yes forgot about that, but indeed it can still be done without it (e.g. `ManimColorInputType = str | int | list[...`)

File: manim/utils/color.py (hunk 1 above)
Comment (MrDiver) [reply manim/utils/color.py]: I fixed that now with just using `ParsableManimColor` as parameter type in `ManimColor`

**Comment by MrDiver** [comment]
Okay i don't want to make this PR more complicated than it needs to be i think all other changes made to the color system should probably be seperate.

⚠️ THIS IS A BREAKING CHANGE ⚠️ 

Just so you with this all references of `colour.py` will be removed and every color should now be an instance of `ManimColor` at some point.

**Comment by MrDiver** [comment]
Still needs HSV support

**Comment by MrDiver** [comment]
Check `color_gradient` 3b1b/manim

**Comment by behackl** [comment]
@MrDiver do you remember what the status of this PR was, exactly?

I think this branch would be a better starting point for further rewrites due to the resolved issue with pickling; perhaps the experimental branch could be rebased on top of this.

**Review by behackl — COMMENTED** [review]
Remaining TODOs, almost done here:

- check remaining TODOs still committed
- fix color example
- perhaps somewhat more information what "included in global namespace" means
- one-line docstrings for `core.py` would be nice

**Comment by MrDiver** [comment]
image
New design for the sphinx directive added in the last commit now using the docutils nodes for easier maintainability

**Comment by MrDiver** [comment]
https://docutils.sourceforge.io/docs/ref/doctree.html#table
For history purposes

File: manim/utils/docbuild/autocolor_directive.py [hunk 2]
```diff
@@ -1,57 +1,98 @@
...
+            row += nodes.entry("", col2)
+        thead += row
+        tgroup += thead
+
+        color_elements = []
+        for member_name, member_obj in inspect.getmembers(module):
+            if isinstance(member_obj, ManimColor):
+                hex_code = member_obj.to_hex()[1:]
+                r, g, b = tuple(int(hex_code[i : i + 2], 16) for i in (0, 2, 4))
+                luminance = (0.2126 * r + 0.7152 * g + 0.0722 * b) / 255
```
Comment (behackl) [inline manim/utils/docbuild/autocolor_directive.py:65]: Can't Python's colorsys do this? Should HSL be supported by ManimColor too?

Comment (MrDiver) [reply manim/utils/docbuild/autocolor_directive.py:65]: Good question, if so then we should change it!

Comment (behackl) [reply manim/utils/docbuild/autocolor_directive.py]: It can't, at least not for the value you compute here. I've simplified the code a bit, regardless.

**Comment by jsonvillanueva** [comment]
For the record, #3264 introduces a new file and also a new class which could use the `ParsableManimColor`. I'm not sure to what extent you're trying to cover all instances of color within the library, but it's shaping out nicely. Should I maybe check this branch out and give it a closer look for/commit other missing occurrences?

File: tests/module/mobject/text/test_text_mobject.py [hunk 3]
```diff
@@ -15,9 +13,10 @@ def test_font_size():
...
-    i.e. colour.Color(red)."""
+# TODO: Figure out if this is still needed
+# def test_non_str_color():
+#     """Test that the Text and MarkupText can accept non_str color values
+#     i.e. colour.Color(red)."""
 
-    text = Text("test_color_inheritance", color=Color("blue"))
-    markup_text = MarkupText("test_color_inheritance", color=Color("blue"))
+#     text = Text("test_color_inheritance", color=Color("blue"))
+#     markup_text = MarkupText("test_color_inheritance", color=Color("blue"))
```
[the remark this answers is not in the document]
Comment (MrDiver) [reply tests/module/mobject/text/test_text_mobject.py]: Done

**Comment by MrDiver** [comment]
> For the record, #3264 introduces a new file and also a new class which could use the `ParsableManimColor`. I'm not sure to what extent you're trying to cover all instances of color within the library, but it's shaping out nicely. Should I maybe check this branch out and give it a closer look for/commit other missing occurrences?

Already ✅ done thank you for the info! that was very helpful to know!

**Comment by MrDiver** [comment]
> Remaining TODOs, almost done here:
> 
>     * check remaining TODOs still committed

There are 2 Todos still remaining, one is the fixing of the color structure passing down colors through constructors and making sure that there are sane default values for everything but it seems to work fine because labeled.py worked flawless without any changes.

The other one with the gradient is just some issue that we kind of also had before, because we just didn't have a good implementation. I just referenced the other function for now. So it can safely be done at a later stage.

>     * fix color example

- [ ] Still needs to be done

>     * perhaps somewhat more information what "included in global namespace" means

In which file do we have this ?

>     * one-line docstrings for `core.py` would be nice

`core.py` is fully documented now!

**Review by behackl — APPROVED** [review]
I am very happy to finally get a sane(r) color model into the library. Thanks to everyone who was involved here, and in particular @MrDiver! :clap:

**Pull request merged**